readme.txt
Introduction
============
This is a small group of Linux utilities for the Atmel SAMA5D2 family of
SoCs (System on a Chip). The SAMA5D2 family could be considered as an
enhancement of the SAMA5D3 and SAMA5D4 families and share many macrocells.
A macrocell is the term used for a named function, or instance of a
function, within the SoC. GPIO lines are grouped in units of 32 and
controlled by macrocells named, PIOA, PIOB, PIOC, and PIOD. Depending on
the context, the more general name "PIO" may be considered as a macrocell.
And that more general sense is how Atmel document the SAMA5D2, with a
section (or chapter) for each in their 2642 page "datasheet".

Some of these utilities work with the Linux kernel while others bypass
the kernel and directly manipulate the macrocell registers which are
memory mapped. Mainline "DT" (device tree) kernels since lk 4.2 have
support for the SAMA5D2 family.

Summary
=======
The following executables are built, their source code can be found in
the src/ directory. Some of these utilities have man pages and they are
found in the doc/ directory.

    a5d2_pio_set     set PIO attributes on given GPIO line
    a5d2_pio_status  fetch PIO status values for given GPIO line
    a5d2_pmc         monitor the Power Management Controller (PMC)
    a5d2_tc_freq     generate squarish waves using the TC macrocell(s)
    devmem2          read or write to memory address (useful for
                     controlling memory mapped IO) N.B. no longer
                     installed since clashes with Debian/Ubuntu
                     package of same name; can be made be hand
    gpio_sysfs       access and test a GPIO line using sysfs interface
    hex2tty          command/response tester for serial lines
    i2c_bbtest       I2C master using user space GPIO bit banging
    i2c_devtest      I2C master using kernel /dev/i2c-<num> device
    is_ariag25       script helper, exit status 0 for at91sam9g25
    is_arm           script helper, exit status 0 for EABI (ARM)
    is_foxg20        script helper, exit status 0 for at91sam9g20
    is_foxglx        script helper, exit status 0 for Axis 384LX
    is_sama5d2       script helper, exit status 0 for sama5d2 family
    is_sama5d3       script helper, exit status 0 for sama5d3 family
    is_sama5d4       script helper, exit status 0 for sama5d4 family
    mem2io           read or write multiple 32 bit values from memory.
    readbits         sysfs based GPIO line reader
    setbits          sysfs based GPIO line writer
    xbee_api         specialization of hex2tty for XBee API

Build system
============
The SAMA5D2 family is powerful enough to run to do the builds itself.
That is simpler than cross building. The following assumes a root
file system based on Debian 7, 8 or something similar.

Connect the system to the internet via one of the ethernet macrocells
(or perhaps a WiFi USB dongle) and make sure the root file system is
up to date:
  # apt-get update
  # apt-get dist-upgrade

To compile and build utilities load the (pseudo) package:
  # apt-get install build-essential

To run the "./build_debian" script to generate a deb binary package
that can be installed on other ARM systems; this pakage is required:
  # apt-get install debhelper

Other "-dev" packages may be needed for some peripherals.

In the top level directory of this package is a script called
build_debian.sh . It will build a Debian binary package and place it in
the directory above the current. It can be invoked like this:
  # ./build_debian.sh

RIf that is not suitable, then this should work:
  # make ; make install

You may need to check where the 'make install' places the executables and
man pages.


Documentation
=============

Read the code (and send corrections to me please). There is usage
information in each utility accessed with the '-h' option. With
some utilities '-hh', '-e' or '-ee' provide extra information (e.g.
a5d2_pio_status and a5d2_pio_set).

gpio_sysfs and gpio_ioctl
=========================
These two utilities have almost identical functionality. They differ in
the way the access the gpio lines at the kernel interface. gpio_sysfs
uses the sysfs gpio interface documented in the
<kernel_source>/Documentation/gpio.txt
file in a section titled "Sysfs Interface for Userspace". gpio_ioctl
uses ioctls on the /dev/gpio device node and is discussed in the next
section.

Both utilities can toggle the specified gpio line <num> times with a
<usec> delay between each transition (see the '-t', '-n' and '-d' options).
The high state may be as a result of an external or internal pull-up
resistor (default action) or may be driven high when the '-f' option
is used. The accuracy of the <usec> delay depends on several factors
(e.g. whether a tickless kernel is used) but it most certainly will
not be accurate down to the microsecond level.

The gpio_sysfs has an extra '-c' option to count rising, falling or
both edges on a gpio line for a given duration (default 1 second).
See the gpio.txt kernel documentation file, specifically about the
"edge" file. The use of poll() from the user space to detect a level
change is quite bizarre and curious readers are encouraged to read the
code for the process_count() function in gpio_sysfs.c .

Recent "device-tree" Linux kernels (e.g. lk 3.7) have changed the
GPIO kernel pin number ordering. In the past it started at 32 (for
PA0) to allow other interrupt number to use the numbers 0 to 31.
Those two numbering systems have been decoupled so GPIO kernel pin
numbers now start at 0 (for PA0). This can be detected at run time
by the presence or absence of this directory: /sys/class/gpio/gpiochip0 .
If it is present PA0 is GPIO kernel pin number 0.

devmem2 and mem2io
==================
The devmem2 program allows memory locations to be read and written.
The helps access the memory-mapped IO locations used by the Atmel
AT91 series of SoCs. It was written by Jan-Derk Bakker
<jdb@lartmaker.nl> in 2000. A copy of that source is in this directory
The author has added a similar program called mem2io which a few extra
features like the ability to read from a script file and change multiple
locations in one invocation.

Both devmem2 and mem2io were built and installed in earlier versions
of this package. However recently a Ubuntu server based system for
ARM was found to have the devmem2 program installed from its own
package. [Debian 6 does not have a devmem2 package in its standard
repositories and devmem2 is flagged as a "requested package" so
it may appear in the future.] So to avoid clashes in the future,
the devmem2 program is no longer built or installed. It can still
be made "by hand" with 'make devmem2'. The mem2io program can be used
as a replacement but it has a slightly different command line syntax.


Douglas Gilbert

dgilbert@interlog.com
21st February 2016


ChangeLog
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. See README file.

Changelog for sama5d2_utils-0.90 [20200502] [svn: r8]
  - setbits: add '-t' and '-T' options to toggle given gpio
  - test basic functionality of a5d2_pio_status, a5d2_pio_set,
    a5d2_pcm and a5d2_tc_freq
  - change Debian compat level from 7 to 10



src/setbits.c
/*
 * Copyright (c) 2010-2020 Douglas Gilbert.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

/*****************************************************************
 * setbits.c
 *
 * Utility for setting a GPIO line on a AT91SAM9G20/25/45 and SAMA5D3/D2 in
 * Linux. Uses sysfs interface. The target hardware is a FoxG20, the Aria
 * G25 and SAMA5D3/D2 family. This utility mimics the actions of a utility of
 * the same name for the FoxLX board. The Aria G25, FoxG20 and FoxLX
 * boards are made by Acme Systems. The SoCs are made by Atmel.
 *
 ****************************************************/

#define _XOPEN_SOURCE 500

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <time.h>
#include <errno.h>


static const char * version_str = "1.09 20200502";

#define EXPORT_FILE "/sys/class/gpio/export"
#define UNEXPORT_FILE "/sys/class/gpio/unexport"
#define GPIO_BASE_FILE "/sys/class/gpio/gpio"
#define PIO_BASE_FILE "/sys/class/gpio/pio"
#define GPIO_BANK_ORIGIN "/sys/class/gpio/gpiochip0"
/* Earlier kernels (G20+G25) offset GPIO numbers by 32 so
 * /sys/class/gpio/gpiochip32 (a directory) was the lowest numbered bank
 * and corresponded to PIOA (therefore PA0 was gpio number 32). By lk 3.7
 * /sys/class/gpio/gpiochip0 existed and corresponded to PIOA (therefore
 * PA0 becomes gpio number 0).
 * In lk 3.8 the sysfs gpio file names changed from /sys/class/gpio/gpio37
 * to /sys/class/gpio/pioB5 however the export and unexport files still
 * expect GPIO numbers.
 */

static int verbose = 0;
static const char * gpio_name = NULL;


static void
usage(void)
{
    fprintf(stderr, "Usage: "
            "setbits [-b BN] [-h] [-p PORT] [-s 0|1] [-S 0|1] [-t] [-T] "
            "[-u] [-U]\n"
            "               [-v] [-V]\n"
            "  where:\n"
            "    -b BN        bit number within a port (0 to 31). Also "
            "accepts\n"
            "                 prefix like 'pc' or just 'c' for PORT (e.g. "
            "'-b c7').\n"
            "    -h           print usage message\n"
            "    -p PORT      port ('a' to 'e') or gpio kernel line number "
            "(0 or more)\n"
            "    -s 0|1       state to set, 0 for low, 1 for high\n"
            "                 (def: set to input, usually with weak "
            "pullup)\n"
            "    -S 0|1       state to set (same action as '-s')\n"
            "    -t           toggle: up(1) then down(0)\n"
            "    -T           toggle: down(0) then up(1)\n"
            "    -u           unexport gpio line prior to setting bit\n"
            "    -U           leave line exported on exit\n"
            "    -v           increase verbosity (multiple times for more)\n"
            "    -V           print version string then exit\n\n"
            "Set GPIO line state on AT91SAM9G20+G25/SAMA5D3+D2 using sysfs. "
            "For Aria G25\nand FoxG20 boards. This utility is similar to a "
            "FoxLX utility of the same\nname. Note: in earlier Linux "
            "kernels gpio kernel line numbers started at\n32 (for PA0), "
            "recent kernels start at 0 (for PA0), use '-vv' to check.\n"
            "Example: 'setbits -b PC7 -s 1'\n");
}

static char *
get_best_gpio_name(int knum, char * b, int blen)
{
    if (gpio_name)
        snprintf(b, blen, "%s [kn=%d]", gpio_name, blen);
    else
        snprintf(b, blen, "knum=%d", knum);
    return b;
}


int
main(int argc, char ** argv)
{
    struct stat sb;
    int opt, k, exp_fd, unexp_fd, dir_fd;
    int bn = -1;
    int knum = -1;
    int state = -1;
    int toggle = 0;      /* 1: (-t) up then down; 2: (-T) down then up */
    int unexport = 0;
    int exported_on_exit = 0;
    int exported = 0;
    int origin0 = 0;
    int ret = -1;
    const char * cp;
    char b[256];
    char base_dir[128];
    char ch;
    char bank = '\0';

    while ((opt = getopt(argc, argv, "b:hp:s:S:tTuUvV")) != -1) {
        switch (opt) {
        case 'b':
            cp = optarg;
            if (isalpha(cp[0])) {
                if ('P' == toupper(*cp))
                    ++cp;
                ch = toupper(*cp);
                if ((ch >= 'A') && (ch <= 'E'))
                    bank = ch;
                else {
                    fprintf(stderr, "'-b' expects a letter ('A' to 'E')\n");
                    exit(EXIT_FAILURE);
                }
                gpio_name = optarg;
                ++cp;
            }
            k = atoi(cp);
            if ((k < 0) || (k > 31)) {
                fprintf(stderr, "'-b' expects a bit number from 0 to 31\n");
                exit(EXIT_FAILURE);
            }
            bn = k;
            break;
        case 'h':
            usage();
            exit(EXIT_SUCCESS);
            break;
        case 'p':
            if (isalpha(*optarg)) {
                ch = toupper(*optarg);
                if ((ch >= 'A') && (ch <= 'E'))
                    bank = ch;
                else {
                    fprintf(stderr, "'-p' expects a letter ('A' to 'E') or "
                            "a number\n");
                    exit(EXIT_FAILURE);
                }
            } else if (isdigit(*optarg)) {
                k = atoi(optarg);
                if ((k < 0) || (k > 511)) {
                    fprintf(stderr, "'-p' expects a letter or a number "
                            "0 or greater\n");
                    exit(EXIT_FAILURE);
                }
                knum = k;
            } else {
                fprintf(stderr, "'-p' expects a letter ('A' to 'E') or "
                        "a number\n");
                exit(EXIT_FAILURE);
            }
            break;
        case 's':
        case 'S':
            state = atoi(optarg);
            if ((state < 0) || (state > 1)) {
                fprintf(stderr, "'-%c' expects '0' or '1'\n", opt);
                exit(EXIT_FAILURE);
            }
            break;
        case 't':
            toggle = 1;
            break;
        case 'T':
            toggle = 2;
            break;
        case 'u':
            ++unexport;
            break;
        case 'U':
            ++exported_on_exit;
            break;
        case 'v':
            ++verbose;
            break;
        case 'V':
            printf("%s\n", version_str);
            exit(EXIT_SUCCESS);
        default: /* '?' */
            usage();
            exit(EXIT_FAILURE);
        }
    }
    if (optind < argc) {
        if (optind < argc) {
            for (; optind < argc; ++optind)
                fprintf(stderr, "Unexpected extra argument: %s\n",
                        argv[optind]);
            usage();
            exit(EXIT_FAILURE);
        }
    }

    if (! ((knum >= 0) || ((bank >= 'A') && (bn >= 0)))) {
        fprintf(stderr, "Expect either '-p PORT' or '-b BN'\n");
        usage();
        exit(EXIT_FAILURE);
    }
    if (stat(GPIO_BANK_ORIGIN, &sb) >= 0) {
        if (verbose > 1)
            fprintf(stderr, "%s found so kernel pin numbers start at 0 "
                    "(for PA0)\n", GPIO_BANK_ORIGIN);
        origin0 = 1;
    } else if (verbose > 2)
        fprintf(stderr, "%s not found so kernel pin numbers start at 32 "
                "(for PA0)\n", GPIO_BANK_ORIGIN);
    if (knum < 0) {
        knum = ((! origin0) + bank - 'A') * 32 + bn;
        if (verbose)
            fprintf(stderr, "%c%d becomes kernel pin number %d\n", bank,
                    bn, knum);
    } else if ((! origin0) && (knum < 32)) {
        fprintf(stderr, "since %s not found assume kernel pin numbers "
                "start at 32\n(for PA0) so %d is too low\n",
                GPIO_BANK_ORIGIN, knum);
        exit(EXIT_FAILURE);
    }
    if ('\0' == bank) {
        bank = (char)((knum / 32) + 'A' - (! origin0));
        bn = knum % 32;
    }

    exp_fd = -1;
    unexp_fd = -1;
    dir_fd = -1;
    exp_fd = open(EXPORT_FILE, O_WRONLY);
    if (exp_fd < 0) {
        perror(EXPORT_FILE);
        goto bad;
    }
    if (unexport || (! exported_on_exit)) {
        unexp_fd = open(UNEXPORT_FILE, O_WRONLY);
        if (unexp_fd < 0) {
            perror(UNEXPORT_FILE);
            goto bad;
        }
    }

    if (unexport && (unexp_fd >= 0)) {
        snprintf(b, sizeof(b), "%d", knum);
        if (pwrite(unexp_fd, b, strlen(b), 0) < 0) {
            fprintf(stderr, "Unable to unexport %s: %s\n",
                    get_best_gpio_name(knum, b, sizeof(b)), strerror(errno));
            fprintf(stderr, "continue ...\n");
        }
    }
    snprintf(b, sizeof(b), "%d", knum);
    if (pwrite(exp_fd, b, strlen(b), 0) < 0) {
        fprintf(stderr, "Unable to export %s (already in use?): %s\n",
                get_best_gpio_name(knum, b, sizeof(b)), strerror(errno));
        goto bad;
    }
    exported = 1;
    /* check if /sys/class/gpio/gpio<knum> directory exists */
    snprintf(base_dir, sizeof(base_dir), "%s%d", GPIO_BASE_FILE, knum);
    if (stat(base_dir, &sb) >= 0) {
        if (verbose > 1)
            fprintf(stderr, "%s found so continue in original manner\n",
                    base_dir);
    } else {
        if (verbose > 2)
            fprintf(stderr, "%s not found, now check for pinctrl "
                    "convention\n", base_dir);
        /* check if /sys/class/gpio/pio<bank><bn> directory exists */
        snprintf(base_dir, sizeof(base_dir), "%s%c%d", PIO_BASE_FILE, bank,
                 bn);
        if (stat(base_dir, &sb) >= 0) {
            if (verbose > 1)
                fprintf(stderr, "%s found so pinctrl convention\n", base_dir);
        } else {
            fprintf(stderr, "Unable to find sysfs directory %s (for "
                    "direction)\n", base_dir);
            goto bad;
        }
    }

    snprintf(b, sizeof(b), "%s/direction", base_dir);
    dir_fd = open(b, O_WRONLY);
    if (dir_fd < 0) {
        fprintf(stderr, "Open %s: %s\n", b, strerror(errno));
        goto bad;
    }
    if ((state < 0) && (0 == toggle)) {
        if (pwrite(dir_fd, "in", 2, 0) < 0) {
            fprintf(stderr, "Unable to write 'in' to dir_fd: %s\n",
                    strerror(errno));
            goto bad;
        }
    } else if (0 == state) {
        if (pwrite(dir_fd, "low", 3, 0) < 0) {
            fprintf(stderr, "Unable to write 'low' to dir_fd: %s\n",
                    strerror(errno));
            goto bad;
        }
    } else if (1 == state) {
        if (pwrite(dir_fd, "high", 4, 0) < 0) {
            fprintf(stderr, "Unable to write 'high' to dir_fd: %s\n",
                    strerror(errno));
            goto bad;
        }
    } else if (1 == toggle) {
        if (pwrite(dir_fd, "high", 4, 0) < 0) {
            fprintf(stderr, "Unable to write 'high' to dir_fd: %s\n",
                    strerror(errno));
            goto bad;
        }
        if (pwrite(dir_fd, "low", 3, 0) < 0) {
            fprintf(stderr, "Unable to write 'low' to dir_fd: %s\n",
                    strerror(errno));
            goto bad;
        }
    } else if (toggle > 1) {
        if (pwrite(dir_fd, "low", 3, 0) < 0) {
            fprintf(stderr, "Unable to write 'low' to dir_fd: %s\n",
                    strerror(errno));
            goto bad;
        }
        if (pwrite(dir_fd, "high", 4, 0) < 0) {
            fprintf(stderr, "Unable to write 'high' to dir_fd: %s\n",
                    strerror(errno));
            goto bad;
        }
    }
    ret = 0;

bad:
    if (dir_fd >= 0)
        close(dir_fd);
    if (unexp_fd >= 0) {
        if (exported && (! exported_on_exit)) {
            snprintf(b, sizeof(b), "%d", knum);
            k = pwrite(unexp_fd, b, strlen(b), 0);
            /* assign to k to suppress warning */
        }
        close(unexp_fd);
    }
    if (exp_fd >= 0)
        close(exp_fd);
    if (ret)
        exit(EXIT_FAILURE);
    else
        return 0;
}
